frontend/src/app/modules/challenges/online-editor/online-editor-page/online-editor-page.component.ts

```diff
@@ -94,16 +85,73 @@ export class OnlineEditorPageComponent implements OnDestroy, OnInit {
         return this.activeTab === title;
     }
 
+    private loadChallenge(challengeId: number) {
+        this.challengeService.getChallengeById(challengeId).subscribe(
+            (challenge) => {
+                this.setupLanguages(challenge);
+                this.setupEditorOptions();
+            },
+            () => {
+                this.router.navigateByUrl('/not-found');
+            },
+        );
+    }
+
+    private setupLanguages(challenge: Challenge) {
+        this.challenge = challenge;
+        this.languages = challenge.versions.map((v) => v.language.name);
+        this.languageVersions = this.extractLanguageVersions(challenge.versions);
+        [this.selectedLanguage] = this.languages;
+        [this.selectedLanguageVersion] = this.languageVersions;
+    }
+
+    private extractLanguageVersions(versions: ChallengeVersion[]) {
+        return versions.reduce((languageVersionNames: string[], version: ChallengeVersion) => {
+            const versionLanguages = version.language.languageVersions.map((languageVersion) => languageVersion.version);
+
+            return languageVersionNames.concat(versionLanguages);
+        }, []);
+    }
```

Comment:
I'd rather do it like this 
```suggestion
        return versions.flatMap((version) =>
            version.language.languageVersions.map((languageVersion) => languageVersion.version));
```